File: BOJ/BOJ_3649_G4_로봇프로젝트.cpp
#define _CRT_SECURE_NO_WARNINGS
#include <iostream>
#include <bits/stdc++.h>


using namespace std;
int arr[1000050];

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);
	cout.tie(NULL);

	int width = 0;

	while (1) {
		int legoCnt;
		cin >> width;
		if (cin.eof())
			break;
		cin >> legoCnt;

		for (int i = 0; i < legoCnt; i++) {
			cin >> arr[i];
		}
		if (legoCnt == 0 || legoCnt == 1) {
			cout << "danger\n";
			continue;
		}
		sort(arr, arr + legoCnt);
		bool isChk = false;
		for (int i = 0; i < legoCnt - 1; i++) {
			int left = i + 1;
			int right = legoCnt - 1;

			int destNum = abs(width * 10000000 - arr[i]);
			while (true) {
				if (left > right)
					break;
				int mid = (left + right) / 2;
				if (arr[mid] == destNum) {
					cout << "yes " << min(arr[i], destNum) << " " << max(arr[i], destNum) << "\n";
					isChk = true;
					break;
				}
				if (destNum < arr[mid]) {
					right = mid - 1;
				}
				else if (arr[mid] < destNum) {
					left = mid + 1;
				}
			}
			if (isChk)
				break;
		}
		if (!isChk)
			cout << "danger\n";
	}
}

File: BOJ/BOJ_3649_G4_로봇프로젝트.java
package BOJ;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.Arrays;
import java.util.Scanner;

public class BOJ_3649_G4_로봇프로젝트 {

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
		StringBuilder sb = new StringBuilder();
		Scanner r = new Scanner(System.in);
		String Swidth;
		while (br.ready()) {
			int width = Integer.parseInt(br.readLine());
			int legoCnt = Integer.parseInt(br.readLine());
			int[] legoArr = new int[legoCnt];
			for (int i = 0; i < legoCnt; i++)
				legoArr[i] = Integer.parseInt(br.readLine());

			if (legoCnt == 0 || legoCnt == 1) {
				sb.append("danger").append("\n");
				continue;
			}
			Arrays.sort(legoArr);
			boolean isChecked = false;
			for (int i = 0; i < legoCnt - 1; i++) {
				int left = i + 1;
				int right = legoCnt - 1;
				int destNum = Math.abs(width * 10000000 - legoArr[i]);
				while (true) {
					if (left > right)
						break;
					int mid = (left + right) / 2;
					if (legoArr[mid] == destNum) {
						sb.append("yes ").append(legoArr[i]).append(" ").append(destNum).append("\n");
						isChecked = true;
						break;
					}
					if (destNum < legoArr[mid])
						right = mid - 1;
					else if (legoArr[mid] < destNum)
						left = mid + 1;
				}
				if (isChecked)
					break;

			}
			if (!isChecked)
				sb.append("danger").append("\n");
		}
		bw.write(sb.toString());
		bw.flush();
	}
}
